processor/resourcedetectionprocessor/internal/aws/eks/detector.go

```diff
@@ -64,160 +61,149 @@ var _ detectorUtils = (*eksDetectorUtils)(nil)
+	d.rb.SetK8sClusterName(apiMeta.ClusterName)
 
-	instanceIdentityDocument, err := imdsClient.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
-	if err != nil {
-		logger.Warn(defaultErrorMessage, zap.Error(err))
-		return ""
+	return d.rb.Emit(), conventions.SchemaURL, nil
+}
+
+func (d *detector) isEKS() (bool, error) {
+	if os.Getenv(kubernetesServiceHostEnvVar) == "" {
+		return false, nil
 	}
 
-	instances, err := ec2Client.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
-		InstanceIds: []string{
-			instanceIdentityDocument.InstanceID,
-		},
-	})
+	clusterVersion, err := d.apiProvider.ClusterVersion()
 	if err != nil {
-		logger.Warn(defaultErrorMessage, zap.Error(err))
-		return ""
+		return false, fmt.Errorf("isEks() error retrieving cluster version: %w", err)
 	}
-
-	clusterName := e.getClusterNameTagFromReservations(instances.Reservations)
-	if len(clusterName) == 0 {
-		logger.Warn("Failed to detect EKS cluster name. No tag for cluster name found on EC2 instance")
-		return ""
+	if strings.Contains(clusterVersion, eksClusterStringIdentifier) {
+		return true, nil
 	}
-
-	return clusterName
+	return false, nil
 }
 
-func (e eksDetectorUtils) getClusterNameTagFromReservations(reservations []types.Reservation) string {
-	for _, reservation := range reservations {
-		for _, instance := range reservation.Instances {
-			for _, tag := range instance.Tags {
-				if tag.Key == nil {
-					continue
-				}
-
-				if *tag.Key == clusterNameAwsEksTag || *tag.Key == clusterNameEksTag {
-					return *tag.Value
-				} else if strings.HasPrefix(*tag.Key, kubernetesClusterNameTag) {
-					return strings.TrimPrefix(*tag.Key, kubernetesClusterNameTag)
-				}
-			}
-		}
+func (e eksDetectorUtils) isIMDSAccessible(ctx context.Context) bool {
+	cfg, err := e.getAWSConfig(ctx, true)
+	if err != nil {
+		e.logger.Debug("EC2 Metadata service is not accessible", zap.Error(err))
+		return false
 	}
 
-	return ""
+	client := imds.NewFromConfig(cfg)
+	if _, err := client.GetMetadata(ctx, &imds.GetMetadataInput{Path: "instance-id"}); err != nil {
+		e.logger.Debug("EC2 Metadata service is not accessible", zap.Error(err))
+		return false
+	}
+	return true
 }
 
-func (e eksDetectorUtils) getCloudAccountID(ctx context.Context, logger *zap.Logger) string {
-	defaultErrorMessage := "Unable to get EKS cluster account ID"
-	cfg, err := config.LoadDefaultConfig(ctx)
+func (e eksDetectorUtils) getAWSConfig(ctx context.Context, checkAccess bool) (aws.Config, error) {
+	awsConfig, err := config.LoadDefaultConfig(ctx)
 	if err != nil {
-		logger.Warn(defaultErrorMessage, zap.Error(err))
-		return ""
+		return aws.Config{}, err
 	}
 
-	imdsClient := imds.NewFromConfig(cfg)
-	instanceIdentityDocument, err := imdsClient.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
-	if err != nil {
-		logger.Warn(defaultErrorMessage, zap.Error(err))
-		return ""
+	maxAttempts := e.cfg.MaxAttempts
+	if checkAccess {
```

Comment:
Is this only for the first check?

Reply:
yes.
This is mainly used by `isIMDSAccessible` and to make sure we only attempt once
I documented this here https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/40205/files#diff-2fd609a720e95f7ad6fe8c31180f20671dbf62301fba3c9914fcb81a8cc930aaR388

Reply:
And if you mean whether to run it "once" , I thought of that and I don't think we should as the state of the network/IMDS might change between detect retries (in case of an error)